bip32.c

```diff
@@ -252,6 +268,194 @@ int hdnode_private_ckd(HDNode *inout, uint32_t i)
 	return 1;
 }
 
+#ifdef USE_CARDANO
+static void multiply8(uint8_t *dst, uint8_t *src, int bytes)
+{
+	int i;
+	uint8_t prev_acc = 0;
+	for (i = 0; i < bytes; i++) {
+		dst[i] = (src[i] << 3) + (prev_acc & 0x8);
+		prev_acc = src[i] >> 5;
+	}
+}
+
+static void add_256bits(uint8_t *dst, uint8_t *src1, uint8_t *src2)
+{
+	int i;
+	for (i = 0; i < 32; i++) {
+		uint8_t a = src1[i];
+		uint8_t b = src2[i];
+		uint16_t r = a + b;
+		dst[i] = r & 0xff;
+	}
+}
+
+int ed25519_scalar_add(const uint8_t *sk1, const uint8_t *sk2, uint8_t *res)
+{
+	bignum256modm s1, s2;
+	expand256_modm(s1, sk1, 32);
+	expand256_modm(s2, sk2, 32);
+	add256_modm(s1, s1, s2);
+	contract256_modm(res, s1);
+	return 0;
+}
+
+int hdnode_private_ckd_cardano(HDNode *inout, uint32_t i)
+{
+	//checks for hardened/non-hardened derivation, keysize 32 means we are dealing with public key and thus non-h, keysize 64 is for private key
+	int keysize = 32;
+	if(i & 0x80000000){
+		keysize = 64;
+	}
+
+	static CONFIDENTIAL uint8_t data[1 + 64 + 4];
+	static CONFIDENTIAL uint8_t I[32 + 32];
+	static CONFIDENTIAL bignum256 a, b;
+	static CONFIDENTIAL uint8_t priv_key[64];
+	static CONFIDENTIAL uint8_t res_key[64];
+
+	memcpy(priv_key,inout->private_key,64);
+
+	if (keysize == 64) { // private derivation
+		data[0] = 0;
+		memcpy(data + 1, inout->private_key, 64);
+	} else { // public derivation
+		hdnode_fill_public_key(inout);
+		data[0] = 2;
+		memcpy(data+1, inout->public_key+1, 32);
+	}
+	write_be(data + keysize + 1, i);
+
+	bn_read_be(inout->private_key, &a);
+
+	static CONFIDENTIAL HMAC_SHA512_CTX ctx;
+	hmac_sha512_Init(&ctx, inout->chain_code, 32);
+	hmac_sha512_Update(&ctx, data, 1 + keysize + 4);
+	hmac_sha512_Final(&ctx, I);
+
+	static CONFIDENTIAL uint8_t zl8[32];
+	multiply8(zl8,I,32);
+	ed25519_scalar_add(zl8,priv_key,res_key);
+	add_256bits(res_key+32,I+32,priv_key+32);
+	memcpy(inout->private_key, res_key, 64);
+
+	if(keysize == 64){
+		data[0]=1;
+	}else{
+		data[0]=3;
+	}
+	hmac_sha512_Init(&ctx, inout->chain_code, 32);
+	hmac_sha512_Update(&ctx, data, 1 + keysize + 4);
+	hmac_sha512_Final(&ctx, I);
+
+	memcpy(inout->chain_code, I + 32, 32);
+	inout->depth++;
+	inout->child_num = i;
+	memzero(inout->public_key, sizeof(inout->public_key));
+
+	// making sure to wipe our memory
+	memzero(&a, sizeof(a));
+	memzero(&b, sizeof(b));
+	memzero(I, sizeof(I));
+	memzero(data, sizeof(data));
+	memzero(priv_key, sizeof(priv_key));
+	memzero(res_key, sizeof(res_key));
+	return 1;
+}
+
+void decitoa(int val, char * out){
+    
+    static char buf[32] = {0};
+    
+    int i = 30;
+    
+    for(; val && i ; --i, val /= 10){
+        buf[i] = "0123456789"[val % 10];
+    }
+    
+    memcpy(out,&buf[i+1],strlen(&buf[i+1])+1);    
+}
+
+int hdnode_from_mnemonic_cardano(HDNode *out, const char *mnemonic){
```

Comment:
Let's change this function (so it matches the API of `hdnode_from_seed`) into:

``` c
int hdnode_from_seed_cardano(const uint8_t *seed, int seed_len, HDNode *out);
```

One will then use the following to get the same functionality:

```
int entropy_len = mnemonic_to_entropy(mnemonic, entropy);
hdnode_from_seed_cardano(entropy, entropy_len / 8, &out);
```